// twilio-chat/lib/media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @classdesc A Media represents a media information for Message in a Channel.
 * @property {String} contentType - content type of media
 * @property {String} sid - The server-assigned unique identifier for Media
 * @property {Number} size - Size of media, bytes
 * @property {String} [filename] - file name if present, null otherwise
 */
class Media {
    constructor(data, services) {
        this.mcsMedia = null;
        this.services = services;
        this.state = {
            sid: data.sid,
            filename: data.filename,
            contentType: data.contentType,
            size: data.size
        };
    }
    get sid() { return this.state.sid; }
    get filename() { return this.state.filename; }
    get contentType() { return this.state.contentType; }
    get size() { return this.state.size; }
    /**
     * Returns direct content URL for the media.
     * @returns {Promise<String>}
     */
    async getContentUrl() {
        if (!this.mcsMedia) {
            if (this.services.mcsClient) {
                this.mcsMedia = await this.services.mcsClient.get(this.state.sid);
            }
            else {
                throw new Error('Media Content Service is unavailable');
            }
        }
        return this.mcsMedia.getContentUrl();
    }
}
exports.Media = Media;
