// twilio-sync/browser/index.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncMapItem = exports.SyncMap = exports.SyncListItem = exports.SyncList = exports.SyncDocument = exports.SyncClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "SyncClient", { enumerable: true, get: function get() {
    return client_1.SyncClient;
  } });
var syncdocument_1 = require("./syncdocument");
Object.defineProperty(exports, "SyncDocument", { enumerable: true, get: function get() {
    return syncdocument_1.SyncDocument;
  } });
var synclist_1 = require("./synclist");
Object.defineProperty(exports, "SyncList", { enumerable: true, get: function get() {
    return synclist_1.SyncList;
  } });
var listitem_1 = require("./listitem");
Object.defineProperty(exports, "SyncListItem", { enumerable: true, get: function get() {
    return listitem_1.ListItem;
  } });
var syncmap_1 = require("./syncmap");
Object.defineProperty(exports, "SyncMap", { enumerable: true, get: function get() {
    return syncmap_1.SyncMap;
  } });
var mapitem_1 = require("./mapitem");
Object.defineProperty(exports, "SyncMapItem", { enumerable: true, get: function get() {
    return mapitem_1.MapItem;
  } });
exports.default = client_1.SyncClient;

module.exports = client_1.SyncClient;
module.exports.SyncClient = client_1.SyncClient;
