// twilio-sync/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncMapItem = exports.SyncMap = exports.SyncListItem = exports.SyncList = exports.SyncDocument = exports.SyncClient = void 0;
const client_1 = require("./client");
Object.defineProperty(exports, "SyncClient", { enumerable: true, get: function () { return client_1.SyncClient; } });
const syncdocument_1 = require("./syncdocument");
Object.defineProperty(exports, "SyncDocument", { enumerable: true, get: function () { return syncdocument_1.SyncDocument; } });
const synclist_1 = require("./synclist");
Object.defineProperty(exports, "SyncList", { enumerable: true, get: function () { return synclist_1.SyncList; } });
const listitem_1 = require("./listitem");
Object.defineProperty(exports, "SyncListItem", { enumerable: true, get: function () { return listitem_1.ListItem; } });
const syncmap_1 = require("./syncmap");
Object.defineProperty(exports, "SyncMap", { enumerable: true, get: function () { return syncmap_1.SyncMap; } });
const mapitem_1 = require("./mapitem");
Object.defineProperty(exports, "SyncMapItem", { enumerable: true, get: function () { return mapitem_1.MapItem; } });
exports.default = client_1.SyncClient;


module.exports = client_1.SyncClient;
module.exports.SyncClient = client_1.SyncClient;
